### PR TITLE
Converts zero-shot transformers to torchimagemodel

### PR DESCRIPTION
**Pull request by manushreegangwar**
## What changes are proposed in this pull request?

Converts zero-shot classification and detection transformer models to `TorchImageModel`.

## How is this patch tested? If it is not, please explain why.

```
import fiftyone as fo
import fiftyone.zoo as foz

dataset = fo.load_dataset("quickstart")
categories = dataset.distinct("ground_truth.detections.label")

model_name = <MODEL_ZOO_NAME>
zoo_model = foz.load_zoo_model(
    model_name,
    classes=categories
)
dataset.apply_model(zoo_model, label_field=<LABEL_FIELD>)

embeddings = dataset.compute_embeddings(model, embeddings_field=<EMB_FIELD>)
```
Results are compared against `develop` (non-TorchImageModel) results.

### Is this a user-facing change that should be mentioned in the release notes?

-   [ ] No. You can skip the rest of this section.
-   [ ] Yes. Give a description of this change to be included in the release
        notes for FiftyOne users.

(Details in 1-2 sentences. You can just refer to another PR with a description
if this PR is part of a larger change.)

### What areas of FiftyOne does this PR affect?

-   [ ] App: FiftyOne application changes
-   [ ] Build: Build and test infrastructure changes
-   [ ] Core: Core `fiftyone` Python library changes
-   [ ] Documentation: FiftyOne documentation changes
-   [ ] Other

## Summary by CodeRabbit

- **Refactor**
  - Improved integration of zero-shot transformer models with enhanced model and device management for greater consistency and reliability.
  - Standardized prediction outputs to return raw logits or processed results, providing clearer and more direct outputs.
  - Unified model configuration and processing logic for a more streamlined experience when working with zero-shot transformer models.

### CODE REVIEW

File: fiftyone/utils/transformers.py [hunk 1]
```diff
@@ -399,8 +403,8 @@ def embed_all(self, args):
     def _embed(self, args):
         inputs = self.processor(images=args, return_tensors="pt")
         with torch.no_grad():
-            image_features = self.model.base_model.get_image_features(
-                **inputs.to(self.device)
+            image_features = self._model.base_model.get_image_features(
```
Comment (jacobsela) [inline fiftyone/utils/transformers.py:406]: What do you think about subclassing from `fout.TorchEmbeddingsMixin`?

File: fiftyone/utils/transformers.py [hunk 2]
```diff
@@ -399,8 +403,8 @@ def embed_all(self, args):
     def _embed(self, args):
         inputs = self.processor(images=args, return_tensors="pt")
```
Comment (jacobsela) [inline fiftyone/utils/transformers.py:404]: This needs to be wrapped in an if baed on the `preprocess` attribute

File: fiftyone/utils/transformers.py [hunk 3]
```diff
@@ -646,29 +609,16 @@ def _predict_from_retrieval(self, arg):
         with torch.no_grad():
             for text_prompt in text_prompts:
                 inputs = self.processor(arg, text_prompt, return_tensors="pt")
```
Comment (jacobsela) [inline fiftyone/utils/transformers.py:611]: same as above needs to be in an if depending on `preprocess`

File: fiftyone/utils/transformers.py [hunk 4]
```diff
@@ -646,29 +609,16 @@ def _predict_from_retrieval(self, arg):
...
-            confidence=conf,
-            logits=logits,
-        )
-
-    def predict_all(self, args):
-        if _has_text_and_image_features(self.model):
-            return self._predict_all_from_features(args)
+    def _forward_pass(self, imgs):
+        if _has_text_and_image_features(self._model):
+            return self._predict_all_from_features(imgs)
```
Comment (jacobsela) [inline fiftyone/utils/transformers.py:619]: `_predict_all_from_features` does more than just the forward pass. `_forward_pass` should just be the actual model computation.

File: fiftyone/utils/transformers.py [hunk 5]
```diff
@@ -787,19 +735,15 @@ def _process_inputs(self, args):
...
-        id2label = {i: c for i, c in enumerate(self.classes)}
-        return to_detections(results, id2label, image_shapes)
+        return results
 
-    def predict(self, arg):
-        target_sizes = [arg.shape[:-1][::-1]]
-        inputs = self._process_inputs(arg)
+    def _forward_pass(self, args):
+        target_sizes = [args.shape[-2:]]
+        inputs = self._process_inputs(args)
```
Comment (jacobsela) [inline fiftyone/utils/transformers.py:746]: Same again here, should be put into preprocess / transforms

File: fiftyone/utils/transformers.py [hunk 6]
```diff
@@ -787,19 +735,15 @@ def _process_inputs(self, args):
 
     def _predict(self, inputs, target_sizes):
         with torch.no_grad():
-            outputs = self.model(**(inputs.to(self.device)))
-
+            outputs = self._model(**(inputs.to(self.device)))
         results = self.processor.image_processor.post_process_object_detection(
```
Comment (jacobsela) [inline fiftyone/utils/transformers.py:739]: This should be in the `_output_processor` attribute

File: fiftyone/zoo/models/manifest-torch.json [hunk 7]
```diff
@@ -2755,7 +2755,9 @@
             "size_bytes": null,
             "default_deployment_config_dict": {
                 "type": "fiftyone.utils.transformers.FiftyOneZeroShotTransformerForImageClassification",
-                "config": {}
+                "config": {
+                    "output_processor_cls": "fiftyone.utils.torch.ClassifierOutputProcessor"
```
Comment (jacobsela) [inline fiftyone/zoo/models/manifest-torch.json:2759]: Can these be made to be the defaults in the config object? Makes it a lot easier to make changes down the line. In general, I would minimize the use of passing default arguments in the manifest. When arguments can come from multiple sources it makes it harder to figure out why things are instantiated the way they are.

**Review by jacobsela — CHANGES_REQUESTED** [review]
